### PR TITLE
[Main] Fix Numerical issue for GTP + Muon + MXFP8 param gather + use layerwise optimizer layout 

### LINKED ISSUE #6390: [feat] GTP+Muon
GTP+Muon related stuffs, including perf optimization, bug-fix, etc..

### PR DESCRIPTION
**Pull request by zhongbozhu**
- [x] I, the PR author, have personally reviewed every line of this PR.

# What does this PR do?

This PR fixed a numerical issue under the specific combination as shown in the title: GTP + Muon + MXFP8 param gather + use layerwise optimizer layout

The fix is simple, let `use_layer_wise_param_layout` for Muon layerwise optimizer respect `reuse_grad_buf_for_mxfp8_param_ag` and choose not to remap parameters into the flat DDP buffer and reuse the gradient buffer for AG. 

This bug also fixed the corner case where Muon + layerwise optimizer + mxfp8 primary weight used to only work with `--overlap-param-gather`, so we need to make sure overlap False also just works. 

**How `_stage_layerwise_mxfp8_params` works for distributed layerwise optimizer for muon**

```
1. Overlap ON

Optimizer step
└─ Update locally owned FP32 masters
   └─ Do not synchronize parameters yet                 # step_with_ready_grads

Next forward:

Bucket group N becomes needed                           # DDP forward pre-hook
├─ If not already dispatched:
│  ├─ Stage every bucket in group N                     # _stage_layerwise_mxfp8_params
│  └─ Launch asynchronous all-gather                    # dist_all_gather_func
├─ Wait for group N                                     # finish_param_sync
├─ Stage and launch group N+1
├─ Publish group N into TE MXFP8 storage                # _post_param_sync
└─ Compute with group N
       └─ overlaps with group N+1 all-gather
Forward pre-hooks and the next-group link progressively synchronize all bucket groups.


2. Overlap OFF

Optimizer step
├─ Update locally owned FP32 masters
└─ Iterate over every LayerWise-owned bucket group      # start_param_sync_for_bucket_group_subset
   ├─ Stage every bucket in the current group           # _stage_layerwise_mxfp8_params
   ├─ Run synchronous all-gather                        # dist_all_gather_func
   ├─ Publish gathered values into TE MXFP8 storage     # _post_param_sync
   └─ Continue with the next bucket group

All groups complete
└─ Optimizer step returns

Both modes synchronize every bucket group; 

overlap ON lets forward pre-hooks drive them incrementally, 

while overlap OFF must traverse all groups eagerly before the next forward begins.
```

Unit test:
```
torchrun --standalone --nproc-per-node=4 \
  --log-dir /tmp/gtp-test-logs \
  --redirects 3 --tee 0:3 \
  -m pytest -q --capture=fd \
  tests/unit_tests/generalized_tensor_parallel/test_gtp_fp8_param_gather.py
```

```
[default0]:PASSED tests/unit_tests/generalized_tensor_parallel/test_gtp_fp8_param_gather.py::TestGTPFp8ParamGather::test_gtp_mxfp8_fp8_param_gather[tp_case0-dp_overlap0]
[default0]:PASSED tests/unit_tests/generalized_tensor_parallel/test_gtp_fp8_param_gather.py::TestGTPFp8ParamGather::test_gtp_mxfp8_fp8_param_gather[tp_case0-dp_overlap1]
[default0]:PASSED tests/unit_tests/generalized_tensor_parallel/test_gtp_fp8_param_gather.py::TestGTPFp8ParamGather::test_gtp_mxfp8_fp8_param_gather[tp_case1-dp_overlap0]
[default0]:PASSED tests/unit_tests/generalized_tensor_parallel/test_gtp_fp8_param_gather.py::TestGTPFp8ParamGather::test_gtp_mxfp8_fp8_param_gather[tp_case1-dp_overlap1]
[default0]:PASSED tests/unit_tests/generalized_tensor_parallel/test_gtp_fp8_param_gather.py::TestGTPFp8ParamGather::test_gtp_mxfp8_moe_fp8_param_gather
[default0]:PASSED tests/unit_tests/generalized_tensor_parallel/test_gtp_fp8_param_gather.py::TestGTPFp8ParamGather::test_gtp_gdp_adamw_moe_muon_mxfp8_sync[False]
[default0]:PASSED tests/unit_tests/generalized_tensor_parallel/test_gtp_fp8_param_gather.py::TestGTPFp8ParamGather::test_gtp_gdp_adamw_moe_muon_mxfp8_sync[True]
[default0]:PASSED tests/unit_tests/generalized_tensor_parallel/test_gtp_fp8_param_gather.py::TestGTPFp8ParamGather::test_muon_layout_and_mxfp8_param_gather_parity[False]
[default0]:PASSED tests/unit_tests/generalized_tensor_parallel/test_gtp_fp8_param_gather.py::TestGTPFp8ParamGather::test_muon_layout_and_mxfp8_param_gather_parity[True]
[default0]:PASSED tests/unit_tests/generalized_tensor_parallel/test_gtp_fp8_param_gather.py::TestGTPFp8ParamGather::test_moe_grouped_tensor_op_fuser_layerwise_mxfp8_sync_parity
[default0]:PASSED tests/unit_tests/generalized_tensor_parallel/test_gtp_fp8_param_gather.py::TestGTPFp8ParamGather::test_gdp_adamw_moe_muon_mxfp8_param_gather[False]
[default0]:PASSED tests/unit_tests/generalized_tensor_parallel/test_gtp_fp8_param_gather.py::TestGTPFp8ParamGather::test_gdp_adamw_moe_muon_mxfp8_param_gather[True]
[default0]:PASSED tests/unit_tests/generalized_tensor_parallel/test_gtp_fp8_param_gather.py::TestGTPFp8ParamGather::test_gtp_mxfp8_save_does_not_perturb_training
```

:warning: For major changes (either in lines of code or in its impact), please make sure to first share a design doc with the team. If you're unsure what's the best way to do so, contact @NVIDIA/mcore-oncall.

## Issue tracking

For PRs from open-source community contributors:

- **New features**: a linked issue is **required**. Please open a [feature request](https://github.com/NVIDIA/Megatron-LM/issues/new?template=feature_request.md) and reference it here before submitting the PR.
- **Small updates (bug fixes, minor improvements)**: a linked issue is **recommended** and will accelerate the PR review process.

Linked issue: 

### Pre-checks

- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

Feel free to message or comment @NVIDIA/mcore-oncall to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

All PRs start as **draft**. If you open a non-draft PR, it will be automatically converted to draft.

#### Step 1: Mark PR as "Ready for Review"

1. When your PR is ready, click **Ready for Review**.
2. An oncall reviewer is auto-assigned and expert reviewers are notified based on your changes.
   - Some PRs may jump straight to step 2. This is determined by `.github/CODEOWNERS`.

:warning: Only mark as ready once merge-conflicts are resolved and the CI is passing.
Final Review might get declined if these requirements are not fulfilled.

#### Step 2: Final Review

For PRs that change `megatron/core`, once all expert reviewers have approved, the `Final Review` label is applied **automatically** and final reviewers are assigned.

For PRs outside `megatron/core`, this step is skipped.

#### Step 3: Approved

Once all required reviewers have approved, the `Approved` label is applied **automatically**.

### Merge

Any member of [mcore-engineers](https://github.com/orgs/NVIDIA/teams/mcore-engineers) will be able to merge your PR.

### CODE REVIEW

File: megatron/core/distributed/param_and_grad_buffer.py [hunk 1]
```diff
@@ -1691,18 +1751,56 @@ def partition_buckets(
...
+        force_single_bucket_group (bool, optional): force all buckets with the same optimizer
+            ownership into a single bucket group.
+        reduce_scatter_with_fp32_accumulation (bool, optional): keep buckets separate when the
+            FP32-accumulating reduce-scatter implementation requires singleton groups.
     """
 
     if len(buffers) == 0:
         return []
 
+    # LayerWiseDistributedOptimizer and DistributedOptimizer classify a whole bucket group from
```
Comment (zhongbozhu) [inline megatron/core/distributed/param_and_grad_buffer.py:1753]: Note: this change is necessary to pass the `--over-param-gather = False` unit test, otherwise the numerical error will be very high. 

With overlap=False, LayerWise launches parameter AG immediately after the Muon step, before Adam fallback has stepped. Without ownership partitioning, the mixed group starts with a Muon bucket, so the entire group—including Adam parameters—is gathered with stale Adam values. DistOpt later skips that group, so updated Adam shards are never gathered.

With overlap=True, AG is deferred to the next forward pass, after both Muon and Adam have stepped. Therefore, the mixed group happens to contain current values.

`partition_buckets` removes this timing dependency by keeping Muon-owned and Adam-owned buckets in separate groups.

File: megatron/core/distributed/param_and_grad_buffer.py [hunk 2]
```diff
@@ -145,6 +145,49 @@ def __init__(
...
+
+            param_slot = local_data_view[
+                param_start - local_shard_start : param_end - local_shard_start
+            ]
+            if param_slot.numel() != main_param.numel():
+                raise RuntimeError(
+                    "LayerWise MXFP8 master and DDP parameter slot must have equal size: "
+                    f"master={main_param.numel()}, slot={param_slot.numel()}."
+                )
+            param_slot.copy_(main_param.detach().reshape(-1))
```
Comment (fanshiqing) [inline megatron/core/distributed/param_and_grad_buffer.py:189]: Is is really safe to perform plat copy for `grouped_mxfp8tensor`?

Comment (fanshiqing) [reply megatron/core/distributed/param_and_grad_buffer.py:189]: please also add  a new UT to cover the case of `moe_use_grouped_tensor` + `use_transformer_engine_op_fuser` + `_stage_layerwise_mxfp8_params` to make sure it's also bitwise match as expected

File: megatron/core/optimizer/optimizer.py [hunk 3]
```diff
@@ -1008,8 +1008,30 @@ def __init__(
...
+                                else None
+                            )
+                            if high_precision_init_val is not None:
+                                main_param = high_precision_init_val.to(
+                                    device=param.device, dtype=torch.float32, copy=True
+                                )
+                                clear_high_precision_init_val = getattr(
+                                    param, 'clear_high_precision_init_val', None
+                                )
+                                if clear_high_precision_init_val is not None:
```
Comment (fanshiqing) [inline megatron/core/optimizer/optimizer.py]: when `high_precision_init_val` is not None, we should always assume `clear_high_precision_init_val` is not None too?

File: megatron/core/optimizer/optimizer.py [hunk 4]
```diff
@@ -1008,8 +1008,30 @@ def __init__(
...
+                                    clear_high_precision_init_val()
+                            else:
+                                main_param = param.detach().clone().float()
```
Comment (fanshiqing) [inline megatron/core/optimizer/optimizer.py:1030]: we can just simplify the above code into something like:

```
if hasattr(param, 'get_high_precision_init_val'):
    main_param = param.get_high_precision_init_val().to(
        device=param.device, dtype=torch.float32, copy=True
    )
    param.clear_high_precision_init_val()
else:
    main_param = param.detach().clone().float()
```

File: megatron/core/distributed/param_and_grad_buffer.py [hunk 5]
```diff
@@ -201,9 +244,14 @@ def __init__(
...
-        # overlap_param_gather covers the layer-wise optimizer case, which sets
-        # overlap_param_gather=True without use_distributed_optimizer.
-        if self.ddp_config.use_distributed_optimizer or self.ddp_config.overlap_param_gather:
+        # LayerWise parameter sync can use this collective group without the distributed
+        # optimizer: asynchronously when overlap is enabled, or synchronously when MXFP8 reuses
+        # grad_data as its BF16 all-gather transport.
+        if (
+            self.ddp_config.use_distributed_optimizer
+            or self.ddp_config.overlap_param_gather
+            or self.ddp_config.reuse_grad_buf_for_mxfp8_param_ag
```
Comment (fanshiqing) [inline megatron/core/distributed/param_and_grad_buffer.py]: nit: cleanup: move these shared 3-way condition check into a single property on ddp instead

**Review by Wohox — APPROVED** [review]
LGTM! 
This PR is highly similar to https://github.com/NVIDIA/Megatron-LM/pull/5479 regarding partition_buckets & high prec init value & copying back to bf16 staging buffer for AG. Discussed with @zhongbozhu #5479 need to rebase current PR and then decide the scope for change.

**Comment by Wohox** [comment]
@zhongbozhu Can you help double check with this PR, the fp8 param gather also utilizes the multi-tensor copy optimization Shiqing developed?

File: megatron/core/distributed/param_and_grad_buffer.py (hunk 2 above)
Comment (zhongbozhu) [reply megatron/core/distributed/param_and_grad_buffer.py:189]: good advice

Comment (zhongbozhu) [reply megatron/core/distributed/param_and_grad_buffer.py:189]: regarding `grouped_mxfp8tensor`, I think it's not really supported  indeed. Has GTP supported single weight? If it's already done, I think I can also handle it here.

**Comment by WanZzzzzz** [comment]
@zhongbozhu What's the impact of this issue? Which benchmark did you observe this issue?

File: megatron/core/optimizer/optimizer.py (hunk 3 above)
Comment (zhongbozhu) [reply megatron/core/optimizer/optimizer.py]: yes

File: megatron/core/optimizer/optimizer.py (hunk 4 above)
Comment (zhongbozhu) [reply megatron/core/optimizer/optimizer.py:1030]: sure

File: megatron/core/distributed/param_and_grad_buffer.py (hunk 5 above)
Comment (zhongbozhu) [reply megatron/core/distributed/param_and_grad_buffer.py]: sure

File: megatron/core/distributed/param_and_grad_buffer.py (hunk 2 above)
Comment (zhongbozhu) [reply megatron/core/distributed/param_and_grad_buffer.py:189]: will ban single weight at config level

**Comment by zhongbozhu** [comment]
> What's the impact of this issue? Which benchmark did you observe this issue?

@WanZzzzzz This only happens in this very combination of Muon + mxfp8 param + GTP + `use_layer_wise_param_layout=True` combination, so it's very specific to nemotron.

File: megatron/core/optimizer/layer_wise_optimizer.py [hunk 6]
```diff
@@ -585,13 +585,27 @@ def __init__(
             if self.ddp_config is not None
             else config.overlap_param_gather
         )
-        if self.overlap_param_gather and not self.use_buffer_param_sync:
```
Comment (zhongbozhu) [inline megatron/core/optimizer/layer_wise_optimizer.py:588]: Note: to fix the numerical bug, this has to be        
```
if not self.use_buffer_param_sync and (
            self.overlap_param_gather or config.reuse_grad_buf_for_mxfp8_param_ag
        ):
```

which is them simplified to be `layerwise_param_sync_via_bucket_group` flag created at init time

File: megatron/core/optimizer/layer_wise_optimizer.py [hunk 7]
```diff
@@ -1005,14 +1019,10 @@ def step_with_ready_grads(self) -> bool:
         # All-gather updated params. If overlap_param_gather is True, the all-gather
         # is deferred to the forward pre-hooks via DDP bucket infrastructure.
         if not self.overlap_param_gather:
-            if self.use_buffer_param_sync:
```
Comment (zhongbozhu) [inline megatron/core/optimizer/layer_wise_optimizer.py:1008]: Note: To fix the numerical bug, this has to be 
```
if self.use_buffer_param_sync or self.config.reuse_grad_buf_for_mxfp8_param_ag:
```

which is then further simplified to be `layerwise_param_sync_via_bucket_group`, essentially if the bucket group should initiate the param sync (instead of letting optimizer step do the param all-gather, and then in this case we don't have all-gather & fwd pass overlap) 

bucket group initiating param sync means that the bucket group chooses the buffer, launches/waits for the collective, and supports forward-hook overlap—instead of the optimizer calling allgather_params() directly, which has no param AG overlap. 

The `reuse_grad_buf_for_mxfp8_param_ag` toggle also requires bucket-group parameter sync because the bucket group owns grad_data and performs the complete FP32 master → BF16 grad buffer → all-gather → TE MXFP8 quant workflow, so it has to be bucket group initiating the param sync, even when `overlap_param_gather` is False.

**Review by fanshiqing — APPROVED** [review]
LGTM

**Review by deepakn94 — APPROVED** [review]

**Review by yaoyu-33 — APPROVED** [review]
